rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,265 @@
+```
+
+## Web Client/API Backward Compatibility Guidelines
+
+A proxy server verison N must support a web client of version N and N-1. A web client of version N _might_ hit a proxy of version 
+N-1. Follow the guidelines below to understand if a new versioned endpoint is needed to support this edge case.
+
+Please follow these guidelines for common scenarios:
+- [Creating a new endpoint](#creating-a-new-endpoint)
+- [Removing an endpoint](#removing-an-endpoint)
+- [Adding/removing a field to a response of an existing endpoint](#addingremoving-a-field-to-a-response-of-an-existing-endpoint)
+- [Updating request body fields](#updating-request-body-fields)
+- [New required fields from an API response in the web client](#new-required-fields-from-an-api-response-in-the-web-client)
+- [Adding new features to the web client](#adding-new-features-to-the-web-client)
+
+You can also use this checklist to make sure you've covered the backward compatibility promise
+### A short checklist when updating the web client/api
+- [ ] If a new endpoint was added to accommodate a new feature, the new feature must fail gracefully with the required proxy version - [more details](#adding-new-features-to-the-web-client)
+- [ ] When updating a feature, check if the web client works as expected with a proxy version N-1
+- [ ] When updating a feature, check that the proxy works receiving requests from web clients N-1
+- [ ] If deprecating an endpoint, a TODO is added to delete in the relevant version - [more details](#removing-an-endpoint)
+- [ ] When updating an API response, deprecated fields are still populated - [more details](#addingremoving-a-field-to-a-response-of-an-existing-endpoint)
+
+
+### Creating a new endpoint
+Creating a new endpoint that previously didn't exist is ok. This endpoint should
+be prefixed with `/v1` if using the REST API paradigm. The response shape should
+always return an object, even if that object would contain only one field.
+Prefer pagination for resources that can be paginated, and always create the
+response as if they _could_ be paginated in the future.
+
+Bad
+```go
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3"}
+	return foos, nil
+}
+```
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    ContainsDuplicates bool
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     ContainsDuplicates: true
+	}, nil
+}
+```
+
+### Removing an endpoint
+An endpoint can be removed in a major version n+2, where n is the last major
+version where the endpoint was used.
+
+Example 1: v17 no longer uses GET /webapi/foos which was last used in v16. The
+endpoint can be removed in v18. 
+Example 2: v4.2.0 no longer uses GET /webapi/bars which was last used in v4.1.3. The endpoint can be removed in v6,
+so that v5 still supports clients using v4.1.3 and v4.2.0.
```

Comment:
it wasn't separated (run on)
```suggestion
Example 1: v17 no longer uses GET /webapi/foos which was last used in v16. The
endpoint can be removed in v18. 

Example 2: v4.2.0 no longer uses GET /webapi/bars which was last used in v4.1.3. The endpoint can be removed in v6, so that v5 still supports clients using v4.1.3 and v4.2.0.
```